### PR TITLE
feat: add LogProcessor to baggagecopy

### PR DESCRIPTION
**Pull request by Jesse0Michael**
add a Log Record Processor to the baggagecopy package that copies members from baggage and adds them to the log record attributes.

---

mimics the `baggagecopy` Span Processor, that copies filtered baggage members from the parent span's context and adds them as span attributes, and instead filters baggage members from the log processor OnEmit context and adds them to the log record as attributes.

followed the prior art of the existing Span Processor closely to include baggage in logs.

### CODE REVIEW

**Comment by dmathieu** [comment]
cc @MikeGoldsmith as code owner.

File: processors/baggagecopy/doc.go [hunk 1]
```diff
@@ -1,27 +1,37 @@
...
+// and adds them as attributes to the span.
 //
 // Keys and values added to Baggage will appear on all subsequent child spans for
 // a trace within this service and will be propagated to external services via
 // propagation headers.
 // If the external services also have a Baggage span processor, the keys and
 // values will appear in those child spans as well.
 //
+// The LogProcessor retrieves [Baggage] from the the context provided when
+// emitting the log dds them as attributes to the log.
```
Comment (pellared) [inline processors/baggagecopy/doc.go:17]:
```suggestion
// emitting the log and adds them as attributes to the log.
```

**Review by pellared — APPROVED** [review]

**Comment by pellared** [comment]
> cc @MikeGoldsmith as code owner.

Also CC @codeboten

File: processors/baggagecopy/log_processor_test.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+func NewTestProcessor() *processor {
+	return &processor{}
+}
+
+func TestLogProcessorAppendsAllBaggageAttributes(t *testing.T) {
+	b, _ := baggage.New()
+	b = addEntryToBaggage(t, b, "baggage.test", "baggage value")
+	ctx := baggage.ContextWithBaggage(context.Background(), b)
+
+	// create log provider with the baggage log processor and a log processor used to capture emitted record
```
Comment (MrAlias) [inline processors/baggagecopy/log_processor_test.go:44]:
```suggestion
```

File: processors/baggagecopy/log_processor_test.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+	wrapped := &processor{}
+	lp := log.NewLoggerProvider(
+		log.WithProcessor(NewLogProcessor(AllowAllMembers)),
+		log.WithProcessor(wrapped),
+	)
+
+	// log a test record
```
Comment (MrAlias) [inline processors/baggagecopy/log_processor_test.go:51]:
```suggestion
```

File: processors/baggagecopy/log_processor_test.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+func TestLogProcessorAppendsBaggageAttributesWithHaPrefixPredicate(t *testing.T) {
+	b, _ := baggage.New()
+	b = addEntryToBaggage(t, b, "baggage.test", "baggage value")
+	ctx := baggage.ContextWithBaggage(context.Background(), b)
+
+	baggageKeyPredicate := func(m baggage.Member) bool {
+		return strings.HasPrefix(m.Key(), "baggage.")
+	}
+
+	// create log provider with the baggage log processor and a log processor used to capture emitted record
```
Comment (MrAlias) [inline processors/baggagecopy/log_processor_test.go:76]:
```suggestion
```

File: processors/baggagecopy/log_processor_test.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+	wrapped := &processor{}
+	lp := log.NewLoggerProvider(
+		log.WithProcessor(NewLogProcessor(baggageKeyPredicate)),
+		log.WithProcessor(wrapped),
+	)
+
+	// log a test record
```
Comment (MrAlias) [inline processors/baggagecopy/log_processor_test.go:83]:
```suggestion
```

File: processors/baggagecopy/log_processor_test.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+	b, _ := baggage.New()
+	b = addEntryToBaggage(t, b, "baggage.test", "baggage value")
+	ctx := baggage.ContextWithBaggage(context.Background(), b)
+
+	expr := regexp.MustCompile(`^baggage\..*`)
+	baggageKeyPredicate := func(m baggage.Member) bool {
+		return expr.MatchString(m.Key())
+	}
+
+	// create log provider with the baggage log processor and a log processor used to capture emitted record
```
Comment (MrAlias) [inline processors/baggagecopy/log_processor_test.go:109]:
```suggestion
```

File: processors/baggagecopy/log_processor_test.go [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+	wrapped := &processor{}
+	lp := log.NewLoggerProvider(
+		log.WithProcessor(NewLogProcessor(baggageKeyPredicate)),
+		log.WithProcessor(wrapped),
+	)
+
+	// log a test record
```
Comment (MrAlias) [inline processors/baggagecopy/log_processor_test.go:116]:
```suggestion
```

File: processors/baggagecopy/log_processor_test.go [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+	b, _ := baggage.New()
+	b = addEntryToBaggage(t, b, "baggage.test", "baggage value")
+	b = addEntryToBaggage(t, b, "foo", "bar")
+	ctx := baggage.ContextWithBaggage(context.Background(), b)
+
+	baggageKeyPredicate := func(m baggage.Member) bool {
+		return m.Key() == "baggage.test"
+	}
+
+	// create log provider with the baggage log processor and a log processor used to capture emitted record
```
Comment (MrAlias) [inline processors/baggagecopy/log_processor_test.go:142]:
```suggestion
```

File: processors/baggagecopy/log_processor_test.go [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+	wrapped := &processor{}
+	lp := log.NewLoggerProvider(
+		log.WithProcessor(NewLogProcessor(baggageKeyPredicate)),
+		log.WithProcessor(wrapped),
+	)
+
+	// log a test record
```
Comment (MrAlias) [inline processors/baggagecopy/log_processor_test.go:149]:
```suggestion
```

File: processors/baggagecopy/log_processor_test.go [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+	var got []api.KeyValue
+	wrapped.records[0].WalkAttributes(func(kv api.KeyValue) bool {
+		got = append(got, kv)
+		return true
+	})
+
+	require.Equal(t, want, got)
+}
+
+func TestLogProcessorAppendsBaggageAttributesWithHaPrefixPredicate(t *testing.T) {
```
Comment (MrAlias) [inline processors/baggagecopy/log_processor_test.go:67]:
```suggestion
func TestLogProcessorAppendsBaggageAttributesWithHasPrefixPredicate(t *testing.T) {
```

File: processors/baggagecopy/log_processor_test.go
Comment (MrAlias) [inline processors/baggagecopy/log_processor_test.go:1]: These tests share a lot of the same code. Can the common parts be refactored into a unified code path?

Comment (Jesse0Michael) [reply processors/baggagecopy/log_processor_test.go:1]: Refactored the tests to be written in a table-driven style.

my original intention was to follow https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/processors/baggagecopy/processor_test.go closely, but maybe this is closer to what you are thinking about. 

LMK if you had something different in mind!

**Review by dmathieu — APPROVED** [review]

**Review by MikeGoldsmith — CHANGES_REQUESTED** [review]
Overall looks goof to me.

I've left one suggestion about setting the default filter in the `NewLogProcessor` instead of `OnEmit`.

File: processors/baggagecopy/log_processor.go [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+	return &LogProcessor{
+		filter: filter,
+	}
+}
+
+// OnEmit adds Baggage member to a log record as attributes that are pulled from
+// the Baggage found in ctx. Baggage members are filtered by the filter passed
+// to NewLogProcessor.
+func (processor LogProcessor) OnEmit(ctx context.Context, record *log.Record) error {
+	filter := processor.filter
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply processors/baggagecopy/log_processor.go:39]: I think we want to guard against zero value use. Having the filter check here, similar to the span processor, seems appropriate.

**Review by MikeGoldsmith — APPROVED** [review]

**Pull request merged**